wallet-new/test/unit/Test/Spec/Keystore.hs

```
@@ -0,0 +1,111 @@
+            Keystore.releaseKeystore keystore
+
+        it "destroying a keystore (completely) works" $ do
+            keystore <- Keystore.newKeystore "test_keystore.key"
+            Keystore.destroyKeystore keystore
+            doesFileExist "test_keystore.key" `shouldReturn` False
+
+        prop "lookup of keys works" $ monadicIO $ do
+            forAllM genKeypair $ \(STB wid, STB esk) -> run $ do
+                withKeystore $ \ks -> do
+                    Keystore.insert wid esk ks
+                    mbKey <- Keystore.lookup wid ks
+                    (fmap hash mbKey) `shouldBe` (Just (hash esk))
+
+        prop "Inserts are persisted after releasing the keystore" $ monadicIO $ do
+            (STB wid, STB esk) <- pick genKeypair
+            run $ do
+                nukeKeystore "test_keystore.key"
+                keystore1 <- Keystore.newKeystore "test_keystore.key"
+                Keystore.insert wid esk keystore1
+                Keystore.releaseKeystore keystore1
+                keystore2 <- Keystore.newKeystore "test_keystore.key"
+                mbKey <- Keystore.lookup wid keystore2
+                (fmap hash mbKey) `shouldBe` (Just (hash esk))
+                -- Let's not forget to tear down the keystore.
+                Keystore.releaseKeystore keystore2
+
+        prop "deletion of keys works" $ monadicIO $ do
+            forAllM genKeypair $ \(STB wid, STB esk) -> run $ do
+                withKeystore $ \ks -> do
+                    Keystore.insert wid esk ks
+                    Keystore.delete wid ks
+                    mbKey <- Keystore.lookup wid ks
+                    (fmap hash mbKey) `shouldBe` Nothing
+
+        prop "Deletion of keys is idempotent" $ monadicIO $ do
+            forAllM genKeypair $ \(STB wid, STB esk) -> run $ do
+                withKeystore $ \ks -> do
+                    Keystore.insert wid esk ks
+                    Keystore.delete wid ks
+                    Keystore.delete wid ks
```

I'm not convinced this is the property we want. I feel that the second call should throw an exception.